### ISSUE TITLE
extract function does not recognise original function's parameter

### ISSUE THREAD

**Issue by benedekfazekas** [clojure-emacs/refactor-nrepl#45]
![extract-fn-bug](https://cloud.githubusercontent.com/assets/1003422/6652548/78a4456e-ca6a-11e4-9d4f-cb557a1e59dd.gif)

**Comment by benedekfazekas** [comment]
there are two problems here:
- the client code does not handle the edge case when you extract function which does not need any parameters
- the above shown example really fails because the I use `name`, shadowing `clojure.core/name` which might be foolish but valid. that results the middleware's find-unbound op returning empty list.

**Comment by benedekfazekas** [comment]
analyzing the current implementation it seems that this is a limitation, not easily resolvable. to fix this we would need a different approach. the draft of this approach:

1. we dig down to the node in the AST representing the s-expression we want to extract
1. we inspect this node's :env :local keys which basically lists all the locals available for our s-expression in its original place
1. we search the subtree representing our s-expression for these locals, if they are used we need them as a parameter in the extracted function. if they are not used we don't need them in the new fn

example, given the original top level expression:
```clojure
(defn mock-author [name]
  (when (= name "paci")
    (let [s-name (str/upper-case name)]
      (generate-string {:name name
                        :s-name s-name
                        :genre "technical"}))))
```

if we want to extract the s-expression starting with "(let " we first run something like:
```clojure
(->> (nth repo-ast 2) ast/nodes (filter #(.startsWith (-> % :form str) "(let")) first :env :locals keys)
;; returns: (name)
```
then something like:
```clojure
(->> (nth repo-ast 2) ast/nodes (filter #(.startsWith (-> % :form str) "(let")) first ast/nodes (filter #(= :local (:op %))) (map :form)) 
;; returns: (name name s-name)
```

that means that we need to have 'name' as a param.

on the other hand if we run the same logic but we want to extract s-expression starting with "(generate-string" our locals will include s-name too so that needs to be param as well.

this solution would simplify the client code (you don't need to extract the s-expression into a top level form before calling the middleware op) and would simplify (most likely ;) ) the middleware code too.

the only tricky bit I see apart from picking the right top level form (already done in find symbols for local-symbols) is to find the s-expression in the AST tree we want to extract. that should be doable tho.

@expez any possible caveats I am missing here? what do you think?

**Comment by expez** [comment]
I wrote about this limitation in the original pull request and considered it a problem not worth fixing.  It's an edge-case and it's considered bad style to shadow functions with variable names so it doesn't happen very often.  A counter-argument is that you might do this inadvertently but with `squiggly-clojure` eastwood will warn you when shadowing stuff like in the example above.

Your approach seems sound, but I think we can get more value by solving other problems for a while longer.  I'm especially thinking about finding and renaming macros, solving classloader isolation and writing some code to rename an entire dir of clj files.

I've also introduced quite a few regressions in `clj-refactor` lately which speaks to the need for more tests.  It might be quite valuable to write some integration tests using `ert`.  These might be fickle on travis, but just running them locally before pushing to `master` might be good enough.
